### PR TITLE
Adds a plugn to extract webpack's manifest into a global

### LINKED ISSUE #42979: [webpack] Do not send the list of files three times
Currently the manifest (webpack runtime + list of files) includes the list of files triplicated, one for each potential extension: `.min.js`, `./min.rtl.css` and `./min.css`.

Looks like two of those tree are added by `@automattic/mini-css-extract-plugin-with-rtl`.

Our file list is huge object of ~22kb. Sending it three times makes the manifest really big (~68kb in total). This script is super critical, because it has to be downloaded and processed before the browser can even request the scripts in the critical path.

A visual guide to truly understand the impact of this:

The image is the **full** manifest. Each read box is a copy of the list of files.
image

### PR DESCRIPTION
**Pull request by scinos**
### Background

Webpack has the concept of [The Manifest](https://webpack.js.org/concepts/manifest/), but it is actually a combination of webpack's runtime and the actual manifest (as in, the list of chunks). Usually this manifest is embedded in the main HTML file, meaning that every single user (even if they have all assets cached) have to download and execute it before it can request the first chunk.

`mini-css-extract-plugin` makes this way worse because it adds another two copies of the list of chunks to the embedded manifest (see #42979)

### Changes

This *experimental* plugin splits the manifest into runtime and the actual manifest (i.e the list of chunks only). This way we can embed the list of chunks (the only part that usually changes between compilations) and extract the runtime to an external file that the browser can cache. This saves some bytes in the critical path.

By extracting the manifest to a global variable it allows `mini-css-extract-plugin` to reuse it, so it doesn't embed the manifest twice again.

Both effects combined should save ~50kb (before gzip) from the critical path for every request.

### How it works

Webpack's manifest has a function like

```
function jsonpScriptSrc(chunkId) {
    return __webpack_require__.p + ({1: '1.js', 2:'2.js'}[chunkId]) +".min.js"
}
```

that transforms a chunkId (eg: `1`) into a file (eg: `1.js`). The map chunkId->file is the manifest, and it can be quite big (>500 entries). This map is composed by webpack by combining the output filename (something like `[name].[chunkhash].min.js`) with the actual list of chunks generated.

This plugin works by hooking into the main template rendering to intercept the variable interpolation in the asset name (eg: `[name].js`). For each var, it will generate a call to a global function (eg: `window.__WEBPACK_MANIFEST.get("name")(chunkId)+".js").

Also, for each var, it will call Webpack variable interpolation logic with just that variable (`getAssetsPath('[name]')), capture the output and save it in a global Map (`window.__WEBPACK_MANIFEST.set('name', function(chunkId){...})`) so it returns the correct variable value for a given `chunkdId`.

Then it will add a second file to the `runtime` chunk with all `window.__WEBPACK_MANIFEST.set()` calls as content. This will allow future improvements were we can inline the manifest (the list of `window.__WEBPACK_MANIFEST.set()` calls) and move the runtime to an external, cacheable script.

So with the above example, we'll end up with:

```
//runtime.js
function jsonpScriptSrc(chunkId) {
    return __webpack_require__.p + window.__WEBPACK_MANIFEST.get("name")(chunkId) +".min.js"
}

//manifest.js
window.__WEBPACK_MANIFEST=new Map();
window.__WEBPACK_MANIFEST.set('name', function(chunkId) {return ({1: '1.js', 2:'2.js'}[chunkId])});
```

The main benefit is that any other plugin (eg: `mini-css-extract-plugin`) needs a way to transform a filename template like `[name].css` it won't need another copy of the full chunkdId->name map. It will reuse the `window.__WEBPACK_MANIFEST.get("name")` call, resulting in a much smaller runtime.

You can see all this in action by loading the live branch and looking for an inline script with the content `__WEBPACK_MANIFEST`. 
 
### Notes

This plugin is right at the limit of my webpack knowledge. I'd like to have some webpack-expert eyes to make sure I didn't misunderstand webpack internals.

### Testing instructions

* Check out the branch and verify `yarn start` still works.
* Load live branch and click around between sections to make sure everything loads properly (including CSS).
* Test it on IE as well.

Fixes #42979

### CODE REVIEW

**Review by sgomes — APPROVED** [review]
LGTM. The code looks good and the changes make sense, with no immediately obvious webpack plugin problems like trying to remember data between hooks. Excellent work, @scinos! 👍 

I tested these changes with WebPageTest, and they do appear to produce a clear improvement, with the page becoming visually complete faster (by 200ms, with the test settings I used) and the request waterfall showing no obvious issues from the extra request being added.

**Comment by nsakaimbo** [comment]
👋 This looks like a really neat optimization! Also as someone for whom webpack is incredibly mysterious, I appreciate the very thorough overview of how the manifest is generated. 

🤔 Seems there's a build error in desktop source ([link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/70823/workflows/ff32f111-f7af-4d5a-bf64-6142adc021c8/jobs/855633)):

```
/home/circleci/wp-calypso/node_modules/webpack/lib/Stats.js:436
						for (const id of chunk.ids) {
						                       ^
TypeError: chunk.ids is not iterable
```

**Comment by jsnajdr** [comment]
When I look at the generated `runtime` chunk, it also contains a build-specific `cssChunks` map with chunk IDs:
```js
/******/                // mini-css-extract-plugin CSS loading
/******/                var cssChunks = {"1":1,"3":1,"4":1,"5":1,"6":1,"7":1,"8":1,"9":1, ... }
```

The map contains chunks that have a CSS asset associated with them, as opposed to chunks that have only a JS file.

If this map is incorrect, two kinds of errors could happen:
- issuing a request for a CSS chunk that doesn't exist, triggering a 404 response and in turn failing the entire dynamic `import()` (which then returns a rejected promise)
- not issuing a request for a CSS chunk even though we should, causing missing CSS and broken UI.

**Comment by scinos** [comment]
> (@nsakaimbo) Seems there's a build error in desktop source

Should be fixed now. I disabled the new plugin for the desktop build, as we don't have an external manifest there.

> (@jsnajdr)  it also contains a build-specific cssChunks map

True. That map is generated by `mini-css-extract-plugin-with-rtl` in https://github.com/Automattic/mini-css-extract-plugin-with-rtl/blob/v0.8.0-with-rtl/src/index.js#L387. Looking at the code it doesn't seem to be affected by the changes in this PR, as it only changes the manifest chunk which doesn't have CSS by definition.

**Comment by jsnajdr** [comment]
> Looking at the code it doesn't seem to be affected by the changes in this PR

My point was that the `runtime` chunk, even after extracting `manifest` (the id:name map) out if it, still contains build-specific information that references chunk IDs. But now I realize that it doesn't break things as I previously thought: if the `cssChunks` value changes, the `runtime` chunk simply gets a new hash and is re-downloaded by the browser. We don't need to extract _all_ the build-specific information from the runtime to make it work.

**Comment by scinos** [comment]
> My point was that the runtime chunk, even after extracting manifest (the id:name map) out if it, still contains build-specific information that references chunk IDs.

Very good point. I'd say for now we keep runtime inlined as well. After all, the biggest benefit of this PR is to avoid triplicating the chunk map, extracting runtime as an external script is a bonus that doesn't add _that_ much.

**Comment by jsnajdr** [comment]
> I'd say for now we keep runtime inlined as well.

I think it'd be interesting to learn how often the `runtime` hash changes after the `manifest` got extracted. My guess is that it will change only rarely, making it a good candidate to _not_ be inline, but rather loading it from a (cached) URL.

The `cssChunks` value will change only when the chunk IDs are reshuffled (adding or removing a chunk, or the chunk splitting algorithm decides to split the shared chunks differently), which doesn't happen that often. On the other hand, to invalidate the `manifest`, a single chunk hash needs to change. That makes caching `manifest` impossible.

Our long term goal should be to make the HTML document as small as possible, extracting everything cache-worthy out of it.

For example, I've been annoyed for a very long time by the rather big `configData` object that we also ship inline in the HTML 🙂

File: client/webpack-utils/extract-manifest-plugin.js
Comment (jsnajdr) [inline client/webpack-utils/extract-manifest-plugin.js:8]: Can we also generate a chunk hash for the `manifest` chunk and add it to the filename?

If the filename is just `manifest.js`, the only good way how to ship it to the browser is to inline it in a `script` tag, because then the filename doesn't matter.

If we want to be flexible, then it should be also possible to load it as URL:
```
<script src="/calypso/evergreen/manifest.bb6fb226f19be0fdc380.min.js"></script>
```

Comment (scinos) [reply client/webpack-utils/extract-manifest-plugin.js:8]: Make sense, let me check if I can get webpack to hash it as well.

**Comment by jsnajdr** [comment]
I think there is a way to make this plugin more robust and elegant.

The purpose of the `assetPath` hook is to replace template variables in a path string like
```
[name].[chunkhash].min.js
```
with the actual values of `name` and `chunkhash`.

The extract manifest plugin's hook tries to detect that exactly this replacement just happened, by looking at `data.hashWithLength` and `data.chunk`. That's a heuristic, I'm not sure how reliable.

Then it extracts the JS expression that's result of replacement of `[name].[chunkhash]`, carefully removing and re-adding the js/css extension, and stores that expression in a global.

The plugin relies on the fact that the templated expression is always in the form of `[name].[chunkhash].min.{js,css}`. If we used anything else, i.e., `[chunkhash]-[name].js`, the results would be incorrect.

I believe there's another way how to organize the code:
- tap into the `assetPath` hook _before_ the `TemplatedPathPlugin` registers its own one
- the plugin will get `[name].[chunkhash].min.js` as its input
- replace the template params we want to override with `window.__WEBPACK_MANIFEST.get( 'name', chunkId )` and `window.__WEBPACK_MANIFEST.get( 'chunkhash', chunkId )` global references
- in the manifest chunk, populate the `__WEBPACK_MANIFEST` value with results of replacement of `[name]` and `[chunkhash]`. These values would come from `TemplatedPathPlugin.hook.assetPath`.

Does that make sense?

**Comment by scinos** [comment]
If the plugin transforms `[name].[chunkhash].min.js` to `${window.__WEBPACK_MANIFEST.get( 'name', chunkId )}.${window.__WEBPACK_MANIFEST.get( 'chunkhash', chunkId )}.min.js`... how does it extract the chunk map outside of the runtime? Wouldn't that require re-implement webpack's logic to replace those variables?

Maybe I'm missing something, I'll try that approach to see if it works.

> The plugin relies on the fact that the templated expression is always in the form of [name].[chunkhash].min.{js,css}. If we used anything else, i.e., [chunkhash]-[name].js, the results would be incorrect.

I'm missing that part. As far as I can tell it relies on the name of the chunk manifest being `runtimeChunk`, and having `.js/.min.js` as the extension of the filename. The only one that relies on the actual name of the file is the assets-writer.

**Comment by jsnajdr** [comment]
> how does it extract the chunk map outside of the runtime? Wouldn't that require re-implement webpack's logic to replace those variables?

One of possible ways is to call `mainTemplate.getAssetPath()` when you need the chunk map:
```js
const nameExpression = compilation.mainTemplate.getAssetPath( JSON.stringify( '[name]' ), {
  chunk: data.chunk, // pass the object with chunk ID maps
  skipExtractPlugin: true, // tell the ExtractManifestPlugin to pass through this call
} );

manifestContent.push( `window.__WEBPACK_MANIFEST.set( 'name', ( chunkId ) => ${ nameExpression } );` );
```

The `skipExtractPlugin` flag is there to break infinite loops (the extract plugin's hook handler calling that very hook and recursively enter itself).

The `__WEBPACK_MANIFEST` can be populated on-demand: only when we encounter a `[path]` template variable do we output a chunk map for `name`.

The whole plugin relies on the fact that when a `chunk` option is passed to the `assetPath` hook, the result of `chunk.getChunkMaps()` is always the same, no matter with which `chunk` we are working with at the moment. These chunk maps are used to construct all the relevant maps, and we can cache them (i.e., store as a global) only when they're constant, of course. I'm not sure if that "fact" is true though 🙂

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@scinos please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/862248). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Comment by jsnajdr** [comment]
> I'm missing that part. As far as I can tell it relies on the name of the chunk manifest being runtime and having `.js/.min.js` as the extension of the filename.

One way to break the current plugin is to set:
```js
outputChunkFilename = '[name].[chunkhash].min.js';
cssChunkFilename = 'style-[name]-[chunkhash].min.css';
```

Then, when the plugin sees a result of replacement by the first pattern:
```js
"" + ({ ...id2names })[chunkId] + "." + ({ ...id2hashes })[chunkId] + ".min.js"
```
it will set the manifest content to:
```js
window.MANIFEST = ( chunkId ) => "" + ({ ...id2names })[chunkId] + "." + ({ ...id2hashes })[chunkId];
```

A moment later the plugin will see a result of replacement by the CSS pattern:
```js
"style-" + ({ ...id2names })[chunkId] + "-" + ({ ...id2hashes })[chunkId] + ".min.css"
```
and replaces the manifest content with:
```js
window.MANIFEST = ( chunkId ) => "style-" + ({ ...id2names })[chunkId] + "-" + ({ ...id2hashes })[chunkId];
```

Then, at runtime, when computing a path for a JS chunk, the manifest will contain code for the CSS pattern, resulting in filename:
```
style-stats-1234abcd.min.js
```
The correct one, though, was:
```
stats.1234abcd.min.js
```

Storing the maps in manifest separately, per replacement variable, solves that:
```js
window.MANIFEST.name = ( chunkId ) => ({ ...id2names })[chunkId];
window.MANIFEST.chunkhash = ( chunkId ) => ({ ...id2hashes })[chunkId];
```

**Comment by jsnajdr** [comment]
> These chunk maps are used to construct all the relevant maps, and we can cache them (i.e., store as a global) only when they're constant, of course. I'm not sure if that "fact" is true though 🙂

There is one scenario where the chunk maps are probably different.

In current Calypso, we generate only _one_ runtime and manifest for all entrypoints.

However, webpack can also be configured to generate _multiple_ runtimes, one for each entrypoint. Even for Calypso, going this way would make a lot of sense. The `login` or `gutenboarding` entrypoints are pretty small and ever load a tiny fraction of the approx 500 chunks that Calypso build produces. Their runtime/manifest therefore doesn't need to contain the whole 500-element map. They only need a tiny subset.

For the `ExtractManifestPlugin` to be generic and universal, it should also support this scenario: be aware there can be multiple runtimes and generate a separate companion manifest for each of them.

I don't think this objection I'm raising now is a blocker for using the `ExtractManifestPlugin` in Calypso. But I think the plugin has a chance to be completely generic and could even be contributed directly to webpack. So, I'm trying to find every fault I can 🙂

File: client/server/bundler/assets-writer.js
Comment (jsnajdr) [inline client/server/bundler/assets-writer.js:55]: This doesn't seem to be true. Maybe it was true at some point, but now, the `runtime` chunk is loaded from a URL (not inlined), while only the `manifest` chunk is inlined.

File: client/server/pages/index.js
Comment (jsnajdr) [inline client/server/pages/index.js:158]: The `manifests` object seems to only ever have one key: `runtime`. Are there scenarios where there are multiple.

Concatenating scripts with a simple newline is not 100% reliable. I've seen some funny examples in the past, and if I remember correctly, they were not theoretical, but happened in practice. Two `<script>` tags next to each other would be safer.

File: client/server/bundler/assets-writer.js
Comment (scinos) [reply client/server/bundler/assets-writer.js:55]: After the last change both are inlined. I think we can tackle moving `runtime` to an external script (if we are confident that the hash will change when the css map will change) in a different PR, to be safer.

**Comment by scinos** [comment]
Ok, I see what you mean. The code is much more robust now indeed. The trick to use webpack itself to get the value for each variable (eg: `[name]`, `[chunckhash]`...) is quite clever!  I really like how it doesn't assume anything about extensions, css or anything else.

However, I still need to do the detection of `! data.hashWithLength || ! ( data.chunk && data.chunk.hashWithLength` to skip the case where the hook is called just to generate a filename (can't generate an expression in that case). Not sure how we can skip that brittle detection.

Relevant generated code:

```
//runtime.<hash>.min.js
function jsonpScriptSrc(chunkId) {
  return __webpack_require__.p + ""+window.__WEBPACK_MANIFEST.get('name')(chunkId)+"."+window.__WEBPACK_MANIFEST.get('chunkhash')(chunkId)+".min.js"
}
...
var href = true && isCssRtlEnabled() ? ""+window.__WEBPACK_MANIFEST.get('name')(chunkId)+"."+window.__WEBPACK_MANIFEST.get('chunkhash')(chunkId)+".min.rtl.css" : ""+window.__WEBPACK_MANIFEST.get('name')(chunkId)+"."+window.__WEBPACK_MANIFEST.get('chunkhash')(chunkId)+".min.css";


//manifest.<hash>.min.js
window.__WEBPACK_MANIFEST=new Map();
window.__WEBPACK_MANIFEST.set('name', (chunkId) => "" + ({"1":"async-load-apps-notifications-index-jsx",...}[chunkId]||chunkId) + ""));
window.__WEBPACK_MANIFEST.set('chunkhash', (chunkId) => "" + ({"1":"adc65d5e27d97f581783",...}[chunkId]||chunkId) + ""));
```

**Comment by nsakaimbo** [comment]
👋 Heads up - looks like this change introduced a runtime error into the desktop app. I'm trying to build and run the app locally to inspect ...

**Comment by jsnajdr** [comment]
> However, I still need to do the detection of `! data.hashWithLength` to skip the case where the hook is called just to generate a filename

From where does that call happen? And how do the arguments look like in this case?

**Comment by scinos** [comment]
> From where does that call happen? And how do the arguments look like in this case?

At least from:

* https://github.com/webpack/webpack/blob/webpack-4/lib/MainTemplate.js#L514 (comes from https://github.com/webpack/webpack/blob/webpack-4/lib/MainTemplate.js#L351)

* https://github.com/webpack/webpack/blob/webpack-4/lib/MainTemplate.js#L521 (comes from https://github.com/webpack/webpack/blob/webpack-4/lib/SetVarMainTemplatePlugin.js#L31)

* https://github.com/webpack/webpack/blob/webpack-4/lib/MainTemplate.js#L527 (comes from https://github.com/webpack/webpack/blob/webpack-4/lib/Compilation.js#L2138)

The second argument is always an object with `hash`, two of them also include a `chunk` and `contentHashType: "javascript"`. 

The "good" one is from https://github.com/webpack/webpack/blob/webpack-4/lib/web/JsonpMainTemplatePlugin.js#L50. The main difference is that the second argument has a `hashWithLength`, and `chunk.hashWithLength` and `chunk. contentHashWithLength` which are not defined in the `Chunk` prototype.

I guess we could also detect that `options.chunk instanceof Chunk` but I don't think that is less brittle. Or maybe search for the characters `[` or `]` in the first arg.

File: client/webpack-utils/extract-manifest-plugin.js
Comment (jsnajdr) [inline client/webpack-utils/extract-manifest-plugin.js:61]: Forwarding _all_ variables, instead of just a curated subset, can be fragile. Also, not sure if forwarding just the `data.chunk` option, instead of the entire `data` object, is 100% correct.

I can break this plugin by using the `[contenthash]` variable in the chunk name template. For `contenthash` to work, we need to forward not only the `chunk` option, but also `contentHashType` (which usually has value `"javascript"`).

Comment (scinos) [reply client/webpack-utils/extract-manifest-plugin.js]: Good point, I changed it to pass the full `data` and `assetInfo` objects to mimic the original function. Also added a try/catch to skip the replacement if there is any problem getting the chunk map replacer function.

Tested it with all allowed variables in Webpack 4 (based on https://github.com/webpack/webpack/blob/webpack-4/lib/TemplatedPathPlugin.js), with a name like `[name]-[chunkhash:11]-[contenthash:3]-[id]-[name]-[query].min.js` and it seems to work. It also works with `[hash]` (but webpack doesn't allow me to use `[hash]` and `[chunkhash]` at the same time, not related to this plugin). Variables `moduleid`, `file`, `modulehash`, `filebase` and `url` also are not allowed by webpack  as `output.filename` or `output.chunkFilename` in our config (looking at the source looks like those unallowed variables are used to generate sourcemap)

I think it is safer to allow passing any variable between `[`,`]` and not only a fixed set. This will make the plugin work even if an external plugin adds a new variable, for example. In theory if `getAssetPath( "...[var]...", options, assetInfo )` works, `getAssetPath("[var]",options, assetInfo)` should always work, independently of what `var` is.

File: client/webpack-utils/extract-manifest-plugin.js
Comment (jsnajdr) [inline client/webpack-utils/extract-manifest-plugin.js:23]: `hock` looks like a typo.

File: client/webpack-utils/extract-manifest-plugin.js
Comment (jsnajdr) [inline client/webpack-utils/extract-manifest-plugin.js:54]: I'd split this condition into two, as they don't have anything in common:
1. Bail out when there's a flag that prevents infinite recursion
2. Bail out when the heuristic detects this is a call that we don't want to intercept.

File: client/webpack-utils/extract-manifest-plugin.js
Comment (jsnajdr) [inline client/webpack-utils/extract-manifest-plugin.js:96]: typo: missing w in ebpack

**Comment by jsnajdr** [comment]
> I guess we could also detect that `options.chunk instanceof Chunk` but I don't think that is less brittle.

I agree that checking for instances of `Chunk` (and excluding these calls) is a pretty good heuristic, but checking for `chunk.hashWithLength` seems to be equally reliable.

The `Chunk` instance is passed when webpack wants to do a simple string replacement in, for example, a chunk file name. There is no JS code involved. The `chunk.id` is simply a string with the ID. And the `Chunk` instance doesn't have a `hashWithLength` field, that's why the heuristic that checks for it works.

We're interested only in `assetPath` calls that replace variables with code fragments. In that case, the caller doesn't pass the `Chunk` instance as is, but creates a custom `chunk` object where, for example, `id` is a `'" + chunkId + "'` JS fragment.

I know that the above is probably obvious to the plugin author, but I had to figure it out on my own when trying to undestand and review the plugin. Maybe parts of the above text could make their way into code comments?

File: client/webpack-utils/extract-manifest-plugin.js
Comment (jsnajdr) [reply client/webpack-utils/extract-manifest-plugin.js]: > Variables `moduleid`, `file`, `modulehash`, `filebase` and `url` also are not allowed by webpack

They are probably related to some exotic library targets, not used for traditional app bundles. Interesting that I couldn't find any usage of the `module*` stuff in the webpack codebase, i.e., any `assetPath` call that would pass `module` as an option.

**Comment by jsnajdr** [comment]
I think there's an issue with the hashes of the `runtime` and `manifest` chunks. I tried the following experiment:
1. Run `build-client-evergreen` on this branch. It will build the Calypso client, together with the `runtime` and `manifest` chunks.
2. Make some small code change that affects only one async-loaded chunks. I modified a string in file `client/blocks/reader-full-post/index.jsx`. Then `build-client-evergreen` again.

**What I expected:** Both builds have an identical `runtime` chunk and their hashes are the same. Only the `manifest` chunks are different, and have different hashes, too. The only thing changed in the `manifest` chunk is the hash value of a single chunk.

**What actually happened:** The contents of the `runtime` chunks are identical, but the two builds assign different hashes to them. That will defeat any attempt to cache the `runtime` chunk. It seems that the `runtime` and `manifest` chunks share a common hash that is computed from both of them. A change in `manifest` therefore invalidates `runtime`, too.

**Comment by jsnajdr** [comment]
At this moment, the only blocker is the incorrect hash of the `runtime` chunk. Other than that, all my comments are about typos or code comments, and the plugin is great otherwise.

**Comment by scinos** [comment]
>  incorrect hash of the runtime chunk

I guess that happens because both files in the `runtime` chunk share the same hash, which is what `[chunkhash]` is expected to do.

A potential alternative would be to not reuse the `runtime` chunk and issue a new `manifest` chunk instead. The problem I found when exploring this approach is that each chunk group can only have one chunk marked with a runtime. I didn't do a thorough investigation of this solution.

Another alternative would be to switch to `[contenthash]` (or `webpack-plugin-hash-output`), where in theory each _file_ will get its own hash based on the content. I'll investigate that option as well, maybe we can use `[contenthash]` only for the manifest file.

**Comment by jsnajdr** [comment]
> Another alternative would be to switch to `[contenthash]`

I think the difference is that if a chunk has two assets, JS and CSS, they each have a unique `contenthash`, while `chunkhash` is for the entire chunk, calculated from both.

Using `contenthash` instead of `chunkhash` sounds like something that deviates from the defaults, or the most common scenario, and is therefore bit risky.

**Comment by scinos** [comment]
> I think the difference is that if a chunk has two assets, JS and CSS, they each have a unique contenthash, while chunkhash is for the entire chunk, calculated from both.

What happens if you change the content of the JS, does the CSS hash change as well? If so, that looks like the same situation with runtime+manifest changes you described above. If that is the case, do you think it is still a blocker?

IMO, runtime and manifest sharing the same hash is not a new problem introduced by this plugin, but just a manifestation of using `[chunkhash]` as part of the filename when a chunk has more than one file. 

I do think we should change it to `[contenthash]` (unsafe as per https://github.com/webpack/webpack/issues/11146) or `webpack-plugin-hash-output` (I prefer this one, but I'm biased), but that something that should be done in a different PR as it will affect lots of files.

**Comment by jsnajdr** [comment]
> What happens if you change the content of the JS, does the CSS hash change as well?

I believe the answer is yes: if you change only JS, the CSS part of the chunk is invalidated, too, as the CSS file gets a new name with a new `[chunkhash]`. Even though the content is identical, the browser will have to download a new file.

> If so, that looks like the same situation with runtime+manifest changes you described above.

The fact that the `runtime` chunk is invalidated on every rebuild, though, defeats all the caching of the runtime that this PR hoped to achieve. Although the `runtime` chunk could be cached for very long periods of time, it isn't.

I'm looking at some options how to give the `runtime` chunk its own `chunkhash`, I'll continue the research tomorrow. If I don't succeed soon, this PR is still mergeable -- it doesn't break things, only doesn't achieve most of the optimizations it hoped for.

**Comment by scinos** [comment]
> The fact that the runtime chunk is invalidated on every rebuild, though, defeats all the caching of the runtime that this PR hoped to achieve.

That would be the ideal scenario, but for now both files (manifest.js and runtime.js) are inlined, so the hash problem won't manifest itself for now.

I'd say we merge this (once I fix the typos) so we can measure the perf benefit of reducing the size of the manifest. Extracting the runtime to a cacheable external script is a different perf improvement that we should do separately.

**Comment by jsnajdr** [comment]
> That would be the ideal scenario, but for now both files (manifest.js and runtime.js) are inlined

I see, so the only impact for now is deduplicating the 3 copies of the maps into one? That will work even without fixing the hashes, that's true.

> I'd say we merge this (once I fix the typos) so we can measure the perf benefit of reducing the size of the manifest.

I'd be suprised if the benefit turned out to have measurable results 🙂 In practice, the HTML response is sent with gzip compression which takes care of the duplicate data very well. Therefore, the transferred size will be approx the same. That leaves only faster parsing and execution. In any case, un-bloating stuff is always great.

**Review by jsnajdr — APPROVED** [review]
I believe this is ready to :ship:. Fix the various typos and nits at your convenience before merging.

**Pull request merged**
